File: clients/js/src/routes/LuckPerms.ts
import type { Player } from "../structs";
import { RouteAdapter } from "./RouteAdapter";

export class LuckPerms extends RouteAdapter {
  /**
   * Adds a permission node to the player
   * @param target Username or UUID of the player to target
   * @param perm Permission node to add
   * @throws If something goes wrong
   */
  public async addPermission(target: string | Player, perm: string) {
    const t = typeof target === "string" ? target : target.uuid;
    await this.client.post(`/luckperms/users/${t}/permissions`, {
      permission: perm,
    });
  }

  /**
   * Removes a permission node from the player
   * @param target Username or UUID of the player to target
   * @param perm Permission node to remove
   * @throws If something goes wrong
   */
  public async removePermission(target: string | Player, perm: string) {
    const t = typeof target === "string" ? target : target.uuid;
    await this.client.delete(`/luckperms/users/${t}/permissions/${perm}`);
  }
}

File: src/main/java/com/tristansmp/elytra/routes/LuckPerms.kt
package com.tristansmp.elytra.routes

import com.tristansmp.elytra.Elytra
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.coroutines.future.await
import net.luckperms.api.node.Node
import java.util.*

data class LuckPermsPermissionsPOST(
    val permission: String,
)

fun Route.LuckPerms() {
    route("/luckperms") {

        delete("/users/{target}/permissions/{permission}") {
            val lp = Elytra.instance.lp

            if (lp == null) {
                call.response.status(HttpStatusCode.FailedDependency)
                call.respond(mapOf("status" to "error", "message" to "LuckPerms not found"))
                return@delete
            }

            val permission = call.parameters["permission"]!!

            val lpUser = lp.userManager.loadUser(UUID.fromString(call.parameters["target"]!!)).await()

            if (lpUser == null) {
                call.response.status(HttpStatusCode.NotFound)
                call.respond(mapOf("status" to "error", "message" to "LuckPerms user not found"))
                return@delete
            }

            lpUser.data().remove(
                Node.builder(permission).build()
            )

            lp.userManager.saveUser(lpUser)

            call.respond(mapOf("status" to "ok"))
        }

        post("/users/{target}/permissions") {
            val req = call.receive<LuckPermsPermissionsPOST>()
            val lp = Elytra.instance.lp

            if (lp == null) {
                call.response.status(HttpStatusCode.FailedDependency)
                call.respond(mapOf("status" to "error", "message" to "LuckPerms not found"))
                return@post
            }

            val lpUser = lp.userManager.loadUser(UUID.fromString(call.parameters["target"]!!)).await()

            if (lpUser == null) {
                call.response.status(HttpStatusCode.NotFound)
                call.respond(mapOf("status" to "error", "message" to "LuckPerms user not found"))
                return@post
            }

            lpUser.data().add(
                Node.builder(req.permission).build()
            )

            lp.userManager.saveUser(lpUser)

            call.respond(mapOf("status" to "ok"))
        }

    }
}

